lib/events/externalcloudaudit/externalcloudaudit.go

```diff
@@ -0,0 +1,420 @@
+	"github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport/api/types/externalcloudaudit"
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/teleport/lib/modules"
+	"github.com/gravitational/teleport/lib/services/local"
+)
+
+const (
+	// TokenLifetime is the lifetime of OIDC tokens used by the
+	// ExternalCloudAudit service with the AWS OIDC integration.
+	TokenLifetime = time.Hour
+
+	refreshBeforeExpirationPeriod = 15 * time.Minute
+	refreshCheckInterval          = 30 * time.Second
+	retrieveTimeout               = 30 * time.Second
+)
+
+// Configurator provides functionality necessary for configuring the External
+// Cloud Audit feature.
+//
+// Specifically:
+//   - IsUsed() reports whether the feature is currently activated and in use.
+//   - GetSpec() provides the current cluster ExternalCloudAuditSpec
+//   - CredentialsProvider() provides AWS credentials for the necessary customer
+//     resources that can be used with aws-sdk-go-v2
+//   - CredentialsProviderSDKV1() provides AWS credentials for the necessary customer
+//     resources that can be used with aws-sdk-go
+//
+// Configurator is a dependency to both the S3 session uploader and the Athena
+// audit logger. They are both initialized before Auth. However, Auth needs to
+// be initialized in order to provide signatures for the OIDC tokens.  That's
+// why SetGenerateOIDCTokenFn() must be called after auth is initialized to inject
+// the OIDC token source dynamically.
+//
+// If auth needs to emit any events during initialization (before
+// SetGenerateOIDCTokenFn is called) that is okay. Events are written to
+// SQS first, credentials from the Configurator are not needed until the batcher
+// reads the events from SQS and tries to write a batch to the customer S3
+// bucket. If the batcher tries to write a batch before the Configurator is
+// initialized and gets an error when trying to retrieve credentials, that's
+// still okay, it will always retry.
+type Configurator struct {
+	// spec is set during initialization of the Configurator. It won't
+	// change, because every change of spec triggers an Auth service reload.
+	spec   *externalcloudaudit.ExternalCloudAuditSpec
+	isUsed bool
+
+	credentialsCache *credentialsCache
+}
+
+// Options holds options for the Configurator.
+type Options struct {
+	clock     clockwork.Clock
+	stsClient stscreds.AssumeRoleWithWebIdentityAPIClient
+}
+
+func (o *Options) setDefaults(ctx context.Context, region string) error {
+	if o.clock == nil {
+		o.clock = clockwork.NewRealClock()
+	}
+	if o.stsClient == nil {
+		var useFips aws.FIPSEndpointState
+		if modules.GetModules().IsBoringBinary() {
+			useFips = aws.FIPSEndpointStateEnabled
+		}
+		cfg, err := config.LoadDefaultConfig(
+			ctx,
+			config.WithRegion(region),
+			config.WithUseFIPSEndpoint(useFips),
+			config.WithRetryMaxAttempts(10),
+		)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		o.stsClient = sts.NewFromConfig(cfg)
```

Comment:
We have lib/cloud/clients.go which can return STS client. Would it make sense to use that so we're consistent in how we're creating clients everywhere? If it doesn't support FIPS endpoint yet, would make sense to add it there as well.

Reply:
I gave this a try but lib/cloud/clients only offers v1 clients, and it's much more difficult to mock the v1 STS client in a way that satisfies https://pkg.go.dev/github.com/aws/aws-sdk-go@v1.47.9/aws/credentials/stscreds#NewWebIdentityRoleProviderWithOptions. Not sure it's worth the effort right now.